### ISSUE TITLE
Link in combination with `repeat[.]` is very inefficient and produces bad PDF

### ISSUE THREAD

**Issue by Holt59** [typst/typst#758]
**Update:** I am updating this issue to propose a minimal reproducible example and to target more appropriately the problem. The original issue is still available below.

The problem occurs when creating `link()` that includes `box()` with `repeat[.]`, typically when creating outline (the default `typst` outline does not include the `...` in the link(), but having a link over the whole line would be way better in my opinion).

Here is the example (this is not a outline, but shows the actual issue):

```typst
#locate(loc => {
  for (i, n) in lorem(20).split().enumerate() {

    // link does not contain box, size is ~43kB
    link(loc, n)
    box(width: 1fr, h(3pt) + box(width: 1fr, repeat[.]) + h(3pt))
    link(loc, str(i))

    // link contains box, size is ~767kB
    link(loc, {
      n
      box(width: 1fr, h(3pt) + box(width: 1fr, repeat[.]) + h(3pt))
      str(i)
    })

    linebreak()
  }
})
```

This code contains two different way of creating an "outline-like" entry, one that only `link()` the title and the number (title / page in outline), and one that link the whole line. When creating a link() over the whole line, the outline becomes almost unusable in the PDF, all the readers I tested have a hard-time displaying the outline and the size of the PDF is ridiculous.

I think the issue is because `link()` will basically create a "PDF link" for every dot, instead of the whole line (this can be checked in the PDF since each dot is clickable independently).

---

**Original issue:**

I cannot show the full document, but I am basically creating a custom outline inspired by the one provided here: https://github.com/typst/typst/issues/488

The generated outlines contain ~70 items of various depth, but using `repeat[.]` or not drastically increase the PDF size and makes my PDF reader (Adobe Acrobat) very slow when looking at the outline.

Basically:

```typst
// PDF size is ~620kB, everything is smooth in adobe
box(width: 1fr, h(3pt) + box(width: 1fr) + h(3pt))

// PDF size is `2450kB, Adobe struggle with the outline
box(width: 1fr, h(3pt) + box(width: 1fr, repeat[.]) + h(3pt))
```

The line is above is **the only line** I change between two compilation.

If I do not link the whole line, the size is also close to the initial one, so I guess the issue is that a link is created for every element in the line (or something like that)?

**Comment by lucifer1004** [comment]
I just met the same problem today and it took me hours to figure out the root cause.

**Comment by cyberbryce** [comment]
For reference: I found that hyperlinks appear to be replicated for values that Typst represents as sequences, as well as those that render across multiple lines in a pdf. E.g.:

1. `#link("https://typst.app/docs/",[Author. _Title_])` will have at least two links for Author and Title because they are text in different text styles, yielding a sequence of content according to `repr`.

2. `#link("https://typst.app/docs/",lorem(50))` is a singleton content but once rendered with the default settings in the web app leads to a pdf with four links, one for each line of text.

**Comment by Andrew15-5** [comment]
Is there any alternative to `repeat` that won't duplicate the annotation link objects, but visually will look the samey?

**Comment by Andrew15-5** [comment]
After a bit of tinkering, I was able to reimplement the `repeat()` element completely that works exactly the same as the `std.repeat()`, but won't produce duplicate link annotations in the PDF file, and therefore won't bloat the document if all the periods in the outline are linked.

https://github.com/typst/typst/issues/5095#issuecomment-2541953110 (possible updates will be there)

<details><summary>repeat implementation in Typst (will not be updated)</summary>


```typ
/// For usage see https://typst.app/docs/reference/layout/repeat/.
///
/// Note: can't detect if the parameter was explicitly specified,  so `auto`
/// is used.
#let repeat(body, gap: auto, justify: auto) = layout(size => {
  let rgap = std.repeat.gap // resolved gap
  let rjustify = std.repeat.justify // resolved justify
  let gap = (if gap != auto { gap } else { rgap }).to-absolute().pt()
  let justify = if justify != auto { justify } else { rjustify }

  let fill = size.width.pt()
  let width = measure(".").width.pt()
  let count = calc.floor((fill + gap) / (width + gap))
  let remaining = calc.rem((fill + gap), (width + gap))

  // crates/typst-library/src/text/lang.rs
  let get-text-dir() = {
    let rtl-langs = (
      "ar",
      "dv",
      "fa",
      "he",
      "ks",
      "pa",
      "ps",
      "sd",
      "ug",
      "ur",
      "yi",
    )
    if text.lang in rtl-langs { rtl } else { ltr }
  }

  let resolve-x-align(align, default: start) = {
    if align.x == start {
      if get-text-dir() == ltr { left } else { right }
    } else if align.x == end {
      if get-text-dir() == ltr { right } else { left }
    } else {
      align.x
    }
  }

  if justify {
    gap += remaining / (count - 1)
  }
  let alignment = std.align.alignment
  let offset = 0
  if count == 1 or not justify {
    let pos = if resolve-x-align(alignment) == left {
      0
    } else if alignment.x == center {
      remaining / 2
    } else {
      remaining
    }
    offset += pos
  }

  // // bug? where text escapes the block/box boundaries.
  // set align(start)
  // box(
  //   // width: fill * 1pt,
  //   width: size.width,
  //   fill: gray.transparentize(50%),
  //   // clip: true,
  //   {
  //     show: align.with(alignment)
  //     box(fill: green,text(tracking: gap * 1pt, range(count).map(_ => ".").join()))
  //   },
  // )

  set align(left) // See above.
  show: move.with(dx: offset * 1pt)
  text(tracking: gap * 1pt, range(count).map(_ => ".").join())
})
```

</details> 

Now you just do this:

```typ
#set outline(fill: repeat[.])
// #set outline(fill: std.repeat[.]) // You can compare the file size difference.
#show outline.entry: it => link(it.element.location(), it)
```

and the file size will _not_ skyrocket for no reason.
